osemgrep: login and logout subcommands

hannesm opened the pull request:
test plan:
  make core
  osemgrep login
  osemgrep logout

Only the login and logout commands are now ported, the api_token is not (yet) used in http requests (this will be done in a followup PR).

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

hannesm committed:
osemgrep: login and logout subcommands

aryx commented on src/osemgrep/cli_login/Login_subcommand.ml:
```diff
@@ -44,19 +44,130 @@ let parse_argv (argv : string array) : conf =
-  *)
-  Exit_code.ok
+  Logs_helpers.setup_logging ~force_color:false ~level:(Some Logs.Debug);
+  Logs.info (fun m -> m "Semgrep version: %s" Version.version);
```
Should probably be Logs.debug.
Does semgrep login --verbose display those things?

hannesm replied:
there's no `--verbose` for login/logout available.

hannesm replied:
removed in [9306c04](https://github.com/returntocorp/semgrep/pull/7637/commits/9306c04c582e7d6724195fea1e0b102f0dc9335a)

aryx commented on src/osemgrep/cli_login/dune:
```diff
@@ -7,6 +7,9 @@
     cmdliner
     commons
 
+    semgrep_core_cli
```
Why do you need semgrep_core_cli?
osemgrep should not depend on semgrep_core_cli because at some point we will remove it
(because there will be no need for having osemgrep and semgrep-core; we would have just osemgrep.

hannesm replied:
for `Version.version` (cli_scan as well uses that) -- but if we don't emit the log message above with version and such, we can just get rid of that :)

hannesm replied:
removed in [9306c04](https://github.com/returntocorp/semgrep/pull/7637/commits/9306c04c582e7d6724195fea1e0b102f0dc9335a)

aryx replied:
yes, just remove those

aryx commented on src/osemgrep/configuring/Semgrep_settings.ml:
```diff
@@ -0,0 +1,97 @@
+    anonymous_user_id = Uuidm.v `V4;
+  }
+
+let _t = ref default_settings
```
why the _prefix? make it sound like it's not used.

hannesm replied:
removed in [9306c04](https://github.com/returntocorp/semgrep/pull/7637/commits/9306c04c582e7d6724195fea1e0b102f0dc9335a)

aryx commented on src/osemgrep/configuring/Semgrep_settings.ml:
```diff
@@ -0,0 +1,97 @@
+type t = {
```
Maybe add a TODO comment like
(* we should use ATD to specify this settings file *)

hannesm replied:
done in [9306c04](https://github.com/returntocorp/semgrep/pull/7637/commits/9306c04c582e7d6724195fea1e0b102f0dc9335a)

aryx commented on src/osemgrep/configuring/Semgrep_settings.ml:
```diff
@@ -0,0 +1,97 @@
+let _t = ref default_settings
+let ( let* ) = Result.bind
+
+let of_yaml = function
```
maybe add a TODO like (* ultimately we should just use ATD to automatically read the settings.yml
file by converting it first to json and then use ATDgen API *)

hannesm replied:
added the comment in [9306c04](https://github.com/returntocorp/semgrep/pull/7637/commits/9306c04c582e7d6724195fea1e0b102f0dc9335a)

aryx commented on src/osemgrep/configuring/Semgrep_settings.ml:
```diff
@@ -0,0 +1,97 @@
+  lazy
+    (_t :=
+       try
+         let _ = File.fullpath settings in
```
what is the line for?

hannesm replied:
checking whether the file exists, otherwise we get a ENOENT from the read_file.

aryx replied:
then maybe better to be explicit and use Sys.file_exists and return a proper error?

hannesm replied:
done in [6ba7cc1](https://github.com/returntocorp/semgrep/pull/7637/commits/6ba7cc17a9567ff72c190dea6a67d7074483904e)

aryx commented on src/osemgrep/configuring/Semgrep_settings.mli:
```diff
@@ -0,0 +1,8 @@
+type t = {
```
Maybe add a comment like
(* content of the the ~/.semgrep/settings.yml file. See also Semgrep_envvars.user_settings_file *)

hannesm replied:
added in [9306c04](https://github.com/returntocorp/semgrep/pull/7637/commits/9306c04c582e7d6724195fea1e0b102f0dc9335a)

aryx commented:
need to add uuidm to semgrep.opam

hannesm committed:
address review comments

hannesm commented:
> need to add uuidm to semgrep.opam

I added that, also rebased on top of current develop.

hannesm commented:
still missing (in contrast to the python version):
- error handling for file write -- here `File.write_file` raises an exception
- checking whether it's an interactive terminal

hannesm committed:
address further review comments:
- use Sys.file_exists explicitly (and also check for a regular file)
- catch errors when writing the settings file

hannesm committed:
Login: check for isatty

hannesm committed:
Add Login_CLI module, allow logging to be configured

aryx commented on src/osemgrep/cli_login/Login_CLI.ml:
```diff
@@ -0,0 +1,114 @@
+(* alt: we could use Logs_cli.level(), but by defining our own flags
+ * we can give better ~doc:. We lose the --verbosity=Level though.
+ *)
+let o_quiet : bool Term.t =
```
We could probably factorize code with Scan_CLI.ml by moving this in CLI_common.ml

aryx replied:
We could also move it in Cmdline_helpers.ml, but the o_verbose doc message seems quite specific to Semgrep,
so maybe better in CLI_common.ml

aryx replied:
I initially hated cmdliner, but I now really like its compostionality! You can compose cmdline args! A bit like dune that is also super compositional; just move your submodule in your repo and boom, it's working. Move around things, no problem still working.

hannesm replied:
agreed, done in [04b56c9](https://github.com/returntocorp/semgrep/pull/7637/commits/04b56c94045713e6539f46d873298cfb71162d21)

hannesm committed:
move logging command line arguments to CLI_common
